### ISSUE TITLE
generic type inference for boolean literals fails

### ISSUE THREAD

**Issue by patsissons** [microsoft/TypeScript#19896]
**TypeScript Version:**  2.6.1
**Code**

```ts
function test<T>(x: T, f: (y: T) => void) { return x; }
let z: false = false;
test(z, x => { z = x; });
// ------------^
// [ts]
// Type 'boolean | undefined' is not assignable to type 'false | undefined'.
// Type 'true' is not assignable to type 'false | undefined'.
```

**Expected behavior:**

`z` and `x` should be of the same type

**Actual behavior:**

`x` is being inferred as `boolean` instead of `false`

**Comment by patsissons** [comment]
I'll try and test this in `@next` when i get a chance, just trying to get this issue posted for now.

**Comment by Proculopsis** [comment]
// let z: false = false;
let z: boolean = false;

Isn't this a compiler bug, letting you use 'false' as a type rather than 'boolean'?

**Comment by patsissons** [comment]
typescript permits boolean literals as type declarators. this is particularly relevant in React development, since `false` means do not render anything, but `true` cannot be rendered.

**Comment by RyanCavanaugh** [comment]
@mhegazy what is the inferred suggestion here?

**Comment by Qwertiy** [comment]
Seems working fine on 3.9.2.

**Comment by jakebailey** [comment]
Just looking at old issues; this was fixed back by #27042.
